### DISCUSSION TITLE
General Plugin Advice

### DISCUSSION THREAD

**Discussion by jranby** [AzuraCast/AzuraCast#5009] [Q&A] [answered]
Hello! 

I asked this over at the Discord channel and got no reply, so trying mu luck here.

I have been getting my head around the plugin functionality and have got stuck around (if even possible?) extending current entities and in turn migrations for the extension, and extending entities to use custom fields like the stationmedia entity. 

Can these things only be done outside of the plugin architecture? If no and possible via plugin is there anything to look at that may help?

Cheers all and great work

**Comment by Vaalyn** [comment, marked as answer]
Hey, sorry I saw your message on Discord but was already heading to bed and forgot to reply there 😓 

I haven't seen anyone else using the plugin system before so I think I might be the only one who has extended some entities / added new ones in the past. This was a few versions ago so I'm not 100% confident in the examples I'm going to put in here but if you encounter a problem with them I'll try to get those sorted out.

In the `events.php` of the plugin add the following listeners and change the `ExamplePlugin` in the paths to the path / namespace of your plugin:

```php
// Add Migrations to the Doctrine migration_paths
$dispatcher->addListener(Event\BuildMigrationConfigurationArray::class, function (Event\BuildMigrationConfigurationArray $event) {
    $migrationConfigurations = $event->getMigrationConfigurations();
    $baseDir = $event->getBaseDir();

    $migrationConfigurations['migrations_paths']['Plugin\ExamplePlugin\Entity\Migration'] = $baseDir . '/plugins/ExamplePlugin/src/Entity/Migration';

    $event->setMigrationConfigurations($migrationConfigurations);
});

// Add Entities to the Doctrine mapping paths
$dispatcher->addListener(Event\BuildDoctrineMappingPaths::class, function (Event\BuildDoctrineMappingPaths $event) {
    $mappingClassesPaths = $event->getMappingClassesPaths();
    $baseDir = $event->getBaseDir();

    $mappingClassesPaths[] = $baseDir . '/plugins/ExamplePlugin/src/Entity';

    $event->setMappingClassesPaths($mappingClassesPaths);
});
```

In your `src/Entity/` directory create a `Migration` directory and a `Repository` directory.

In the `Entity` directory you can create your new entities or extend existing ones like this:

```php
<?php

namespace Plugin\ExamplePlugin\Entity;

use App\Entity\Attributes;
use App\Entity\StationPlaylist as AzuraCastStationPlaylist;
use Doctrine\ORM\Mapping as ORM;

#[
    ORM\Entity,
    ORM\Table(name: 'station_playlists'),
    ORM\HasLifecycleCallbacks,
    Attributes\Auditable
]
class StationPlaylist extends AzuraCastStationPlaylist
{
}
```

To generate the migration file from the entities you need to run `./docker.sh cli migrations:diff`.
There is one issue with the generation of those files which we haven't figured out yet. The migration files will be generated in the `src/Entities/Migrations` directory of the AzuraCast dev installation instead of the one of the plugin. You'll have to manually move the file over to the plugins directory after generating it (and check if the namespace was generated correctly too, can't remember if that one was okay or not).

Let me know if you have any questions or issues.

**Reply by jranby** [reply]
That is just what I was after, gives plenty to go on.

I will report back and share what I manage to do.

Very much appreciated, thanks.

**Comment by jranby** [comment]
@Vaalyn I am getting fatal errors regarding the $dispatcher type passed in by the example plugin?

```
Fatal error: Uncaught TypeError: App\Plugins::{closure}(): Argument #1 ($dispatcher) must be of type App\EventDispatcher, Azura\SlimCallableEventDispatcher\SlimCallableEventDispatcher given in /var/azuracast/www/plugins/example-plugin/events.php on line 7

TypeError: App\Plugins::{closure}(): Argument #1 ($dispatcher) must be of type App\EventDispatcher, Azura\SlimCallableEventDispatcher\SlimCallableEventDispatcher given in /var/azuracast/www/plugins/example-plugin/events.php on line 7
```

I am running PHP 8.1 FYI. Any ideas?

**Reply by BusterNeece** [reply]
I think our sample plugin needs updating to be compatible with some minor type changes we've implemented.

I haven't really done much with our plugin system lately because of its low level of use (and the relative difficulty it involves in integrating it into the application).

**Reply by jranby** [reply]
@SlvrEagle23 surprises me that it hasn't seen more use tbh. What is the likelihood of an update do you think? Appreciate that it is not a priority, and thanks for taking the time to reply.

**Reply by Vaalyn** [reply]
@jranby I have started updating the example plugin for AC 0.15 compatibility.

Here is the draft PR with the current progress: https://github.com/AzuraCast/example-plugin/pull/5/files

Please try adapting your `events.php` to use the `CallableEventDispatcherInterface` (with the `use` declaration as seen in the PR) instead of the old `\App\EventDispatcher`
